crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -209,17 +252,59 @@ pub(crate) async fn collect_session_recipients(
     Ok(CollectRecipientsResult { should_rotate, devices, withheld_devices })
 }
 
+// Checks if the session has been shared with a device that is not anymore in
+// the pool of devices that should participate in the discussion.
+fn is_session_overshared_for_user(
+    outbound: &OutboundGroupSession,
+    user_id: &UserId,
+    recipients: &[DeviceData],
+) -> bool {
+    // Device IDs that should receive this session
+    let recipient_device_ids: BTreeSet<&DeviceId> =
+        recipients.iter().map(|d| d.device_id()).collect();
+
+    if let Some(shared) = outbound.shared_with_set.read().unwrap().get(user_id) {
+        // Devices that received this session
+        let shared: BTreeSet<OwnedDeviceId> = shared.keys().cloned().collect();
+        let shared: BTreeSet<&DeviceId> = shared.iter().map(|d| d.as_ref()).collect();
+
+        // The set difference between
+        //
+        // 1. Devices that had previously received the session, and
+        // 2. Devices that would now receive the session
+        //
+        // Represents newly deleted or blacklisted devices. If this
+        // set is non-empty, we must rotate.
+        let newly_deleted_or_blacklisted =
+            shared.difference(&recipient_device_ids).collect::<BTreeSet<_>>();
+
+        let should_rotate = !newly_deleted_or_blacklisted.is_empty();
+        if should_rotate {
+            debug!(
+                "Rotating a room key due to these devices being deleted/blacklisted {:?}",
+                newly_deleted_or_blacklisted,
+            );
+        }
+        should_rotate
+    } else {
+        false
+    }
+}
+
 struct RecipientDevices {
     allowed_devices: Vec<DeviceData>,
     denied_devices_with_code: Vec<(DeviceData, WithheldCode)>,
+    unsigned_of_verified_user: Option<Vec<OwnedDeviceId>>,
```

Comment:
Initially the 2 sharing strategies were using the same structure to return what devices should or should not receive the key for a given user.
But now with the new option we are adding, the strategies differs on what they do. So this `unsigned_of_verified_user` is specific to device based strategy but this struct is used by both strategy. I put it as an option for now.
I try another alternative that is to return a `Result<RecipientDevices>` instead, then each strategy could do their custom errors.
But then the enclosing code had to aggregate the Error per user into a more global error (to gather all the errors). It seemed a bit of an overkill? so for now going for that. But open to suggestions (given it's going to get worse with more options)